Fetch all jobs on a `gh view run` with `json` flag

harveysanders opened the pull request:
Fixes #7771 
Fetches all of a run's jobs across all pages.
#### Ex:
The following run has 60 jobs. Before this feature, `run view [runID] --json` would only fetch up to 30 jobs. Now it will fetch all of a run's jobs.
https://github.com/harveysanders/WhatTheHack/actions/runs/5771576940
```
$ gh run view 5771576940 --json "jobs" | jq '.jobs | length'
60
```

harveysanders committed:
Fetch all jobs on a gh view run --json

andyfeller commented on pkg/cmd/run/shared/shared.go:
```diff
@@ -424,18 +424,32 @@ func preloadWorkflowNames(client *api.Client, repo ghrepo.Interface, runs []Run)
 	var result JobsPayload
-	if err := client.REST(repo.RepoHost(), "GET", run.JobsURL, nil, &result); err != nil {
-		return nil, err
+	perPage := 100
+	v := url.Values{}
+	v.Set("per_page", fmt.Sprintf("%d", perPage))
+
+	// Fetch all the pages until total fetched jobs == result.TotalCount
+	maxPages := 10
+	for p := 1; p < maxPages; p++ {
+		v.Set("page", fmt.Sprintf("%d", p))
+		jobsPath := fmt.Sprintf("%s?%s", run.JobsURL, v.Encode())
+		if err := client.REST(repo.RepoHost(), "GET", jobsPath, nil, &result); err != nil {
+			return run.Jobs, err
+		}
+		run.Jobs = append(run.Jobs, result.Jobs...)
+		if len(run.Jobs) >= result.TotalCount {
+			break
+		}
```
This should be rewritten to take advantage of [`(c Client) RESTWithNext(hostname string, method string, p string, body io.Reader, data interface{}) (string, error)`](https://github.com/cli/cli/blob/bb42fa07aa7fd6044a875a82061ec3fe073a4962/api/client.go#L112) as this should handle the pagination and aggregation 👍 

For [`client.RESTWithNext()` examples](https://github.com/search?q=repo%3Acli%2Fcli%20%22client.RESTWithNext%22&type=code)

harveysanders replied:
Ah, thanks for the suggestion. I'll take a look and refactor it : )

andyfeller commented on pkg/cmd/run/view/view_test.go:
```diff
@@ -1242,6 +1242,51 @@ func TestViewRun(t *testing.T) {
+					[]string{"jobs"},
+					func(io *iostreams.IOStreams, data interface{}) error {
+						run, ok := data.(*shared.Run)
+						if !ok {
+							return fmt.Errorf("expected data type *shared.Run")
+						}
+						fmt.Fprintf(io.Out, "fetched %d jobs\n", len(run.Jobs))
+						return nil
+					},
+				),
+			},
+			httpStubs: func(reg *httpmock.Registry) {
+				jobs := []shared.Job{}
+				for i := 0; i < 115; i++ {
+					jobs = append(jobs, shared.Job{
+						ID:   int64(i),
+						Name: fmt.Sprintf("job %d", i),
+					})
+				}
+
+				limit := 100
+				firstPage := jobs[:limit]
+				secondPage := jobs[limit:]
+
+				reg.Register(
+					httpmock.REST("GET", "repos/OWNER/REPO/actions/runs/3"),
+					httpmock.JSONResponse(shared.SuccessfulRun))
+				reg.Register(
+					httpmock.REST("GET", "repos/OWNER/REPO/actions/workflows/123"),
+					httpmock.JSONResponse(shared.TestWorkflow))
+				reg.Register(
+					httpmock.REST("GET", "runs/3/jobs"),
+					httpmock.JSONResponse(shared.JobsPayload{TotalCount: len(jobs), Jobs: firstPage}))
+				// /runs/:runID/jobs endpoint should be called once more to fetch the remaining jobs
+				reg.Register(
+					httpmock.REST("GET", "runs/3/jobs"),
+					httpmock.JSONResponse(shared.JobsPayload{TotalCount: len(jobs), Jobs: secondPage}))
+			},
+			wantOut: "fetched 115 jobs\n",
+		},
```
In rewriting the implementation with the pagination function, I was wondering how some of those other scenarios did testing versus this.

What do you think about the following?

https://github.com/cli/cli/blob/bb42fa07aa7fd6044a875a82061ec3fe073a4962/pkg/cmd/secret/list/list_test.go#L392-L410

https://github.com/cli/cli/blob/bb42fa07aa7fd6044a875a82061ec3fe073a4962/pkg/cmd/variable/list/list_test.go#L257-L275

harveysanders replied:
Yes, something like this should work for the view tests as well. These cases are probably clearer to the reader at first glance and use a bit less logic/setup. I'll take a closer look at these examples and rewrite my tests.
Thanks!

andyfeller reviewed (changes requested):
@harveysanders : Thank you for opening up this PR and contributing! ✨ 

I applaud you for running into this use case; would love hearing more about a workflow with more than the default items returned!  I left a few suggestions because I think there are some easier ways to implement it but the spirit of what you're trying to do makes sense.

harveysanders commented:
@andyfeller Thanks for reviewing my PR!
I personally have never run into this issue as I usually have much fewer than 30 jobs in my workflows. I just thought [this issue](https://github.com/cli/cli/issues/7771) would be a fun one to pick up. I would be interested in knowing how many devs run into this issue.

100terres commented:
> I would be interested in knowing how many devs run into this issue.

@harveysanders We are facing the issue. On the project I'm working on, we tend to have ~50-60 jobs for our CI workflow. We've recently noticed this limitation with `gh` since only 30 jobs we're reported on monitoring tool. _(we are using `gh` to gather some of the metrics we monitor)_

Thank you for taking the time to fix this issue! 👏

harveysanders committed:
Merge branch 'trunk' of github.com:cli/cli into 7771-list-all-run-jobs

* 'trunk' of github.com:cli/cli:
  Add JSON output flags to workflow list command (#7902)
  Add config clear-cache command (#7693)
  Allow --disable-wiki flag with --template flag in repo create command (#7886)
  Reword token invalid message (#7906)
  🐛 (repo create)execute clone when readme option is enabled (#7837)
  Use go-gh prompter package (#7896)
  pr checks return distinct exit code for PENDING checks (#7866)
  Properly rename remote for GHES hosts (#7890)
  Add verbose flag to api cmd (#7826)
  Fix up Nix installation instructions (#7891)

harveysanders committed:
Refactor GetJobs to use RESTWithNext

harveysanders committed:
Fix pagination resp with "Link" header

harveysanders committed:
Use "Link" header for pagination completion check

harveysanders committed:
Simplify mock jobs response

harveysanders commented:
@andyfeller I've finally got some time to update the PR. I've switched over to `client.RESTWithNext()` and have updated the tests to respond with the proper `"Link"` pagination header. I've also simplified the mock responses while still ensuring all the jobs are collected from each request. 
Please give it another look when you get a chance. Thanks!

harveysanders committed:
Merge branch 'trunk' into 7771-list-all-run-jobs

andyfeller commented on pkg/cmd/run/shared/shared.go:
```diff
@@ -424,19 +425,31 @@ func preloadWorkflowNames(client *api.Client, repo ghrepo.Interface, runs []Run)
+		var err error
+		jobsPath, err = client.RESTWithNext(repo.RepoHost(), http.MethodGet, jobsPath, nil, &resp)
+		if err != nil {
+			return run.Jobs, err
```
Similar `cli/cli` code returns nothing if any error is raised like the original code.

```suggestion
			return nil, err
```

harveysanders replied:
@andyfeller Got it. Along those same lines, does it make sense to reset `run.Jobs` back to `nil` if there is an error?
https://github.com/harveysanders/cli/blob/e96fe1d007f401db6660dcf250d33c69a7c8ae8d/pkg/cmd/run/shared/shared.go#L433-L435

Or would you prefer to use an internal `Job`s slice and only update `run.Jobs` on successful completion?
```go
var results []Job
// fetch jobs
// ....
results = append(results, resp.Jobs...)
// ....
// all fetching complete
run.Jobs = results
return run.Jobs, err
```

harveysanders replied:
I opted for the former
https://github.com/cli/cli/pull/7879/commits/6f0f5a89e2538c20cb02b3e1951026f08ea1e082

andyfeller commented on pkg/cmd/run/shared/shared.go:
```diff
@@ -424,19 +425,31 @@ func preloadWorkflowNames(client *api.Client, repo ghrepo.Interface, runs []Run)
-	if err := client.REST(repo.RepoHost(), "GET", run.JobsURL, nil, &result); err != nil {
-		return nil, err
+
+	v := url.Values{}
+	v.Set("per_page", "100")
```
This is a minor nit but I thought it important to keep similar variable naming as other code:

```suggestion
	query := url.Values{}
	query.Set("per_page", "100")
```

harveysanders replied:
Good call. Thanks for pointing this out :)

andyfeller commented on pkg/cmd/run/shared/shared.go:
```diff
@@ -424,19 +425,31 @@ func preloadWorkflowNames(client *api.Client, repo ghrepo.Interface, runs []Run)
+	v := url.Values{}
+	v.Set("per_page", "100")
+
+	jobsPath := fmt.Sprintf("%s?%s", run.JobsURL, v.Encode())
```
```suggestion
	jobsPath := fmt.Sprintf("%s?%s", run.JobsURL, query.Encode())
```

andyfeller reviewed (changes requested):
@harveysanders : These changes are looking much better with 1 nit (🙏) and 1 behavior change requested.  Other than these, I think the review is shippable afterwards, which I'll keep 👀 on.

Thank you again for your patience and time ❤️

harveysanders committed:
Use consistent variable naming

harveysanders commented:
@andyfeller Thanks again for the review! I honestly appreciate the nits. I've updated the PR with the changes

harveysanders committed:
Merge branch 'trunk' into 7771-list-all-run-jobs

andyfeller reviewed (approved):
@harveysanders : :fishsticks: and ✨ this looks ready, :shipit:

andyfeller commented:
@harveysanders : we typically do biweekly releases and are just about due for the next one, you should receive a notification from the mention in the release notes
